### ISSUE TITLE
Kubelet restarts due to concurrent map read and write

### ISSUE THREAD

**Issue by sseetharaman6** [kubernetes/kubernetes#99370]
#### What happened:
Kubelet restarted due to `fatal error: concurrent map read and map write`.  Topology manager enabled with best-effort policy, CPU manger is not running. 
```
Feb 22 05:15:41 node-01 kubelet[2837]: I0222 05:15:41.093989    2837 topology_manager.go:258] [topologymanager] Topology Affinity for (pod: c44b21d4-bd8b-4183-989e-420810978ed9 container: compute): {01 true}
Feb 22 05:15:41 node-01 kubelet[2837]: I0222 05:15:41.096085    2837 endpoint.go:112] State pushed for device plugin nvidia.com/1e37
Feb 22 05:15:41 node-01 kubelet[2837]: fatal error: concurrent map read and map write
Feb 22 05:15:41 node-01 kubelet[2837]: goroutine 286 [running]:
Feb 22 05:15:41 node-01 kubelet[2837]: runtime.throw(0x4656afe, 0x21)
Feb 22 05:15:41 node-01 kubelet[2837]:         /usr/local/go/src/runtime/panic.go:1116 +0x72 fp=0xc0005cafe8 sp=0xc0005cafb8 pc=0x43b032
Feb 22 05:15:41 node-01 kubelet[2837]: runtime.mapaccess2_faststr(0x3d01520, 0xc0009c7650, 0xc002416c60, 0x18, 0xc0024c6988, 0xc002236000)
Feb 22 05:15:41 node-01 kubelet[2837]:         /usr/local/go/src/runtime/map_faststr.go:116 +0x4a5 fp=0xc0005cb058 sp=0xc0005cafe8 pc=0x417c45
Feb 22 05:15:41 node-01 kubelet[2837]: k8s.io/kubernetes/pkg/kubelet/cm/devicemanager.(*ManagerImpl).isDevicePluginResource(...)
Feb 22 05:15:41 node-01 kubelet[2837]:         /workspace/anago-v1.19.2-rc.0.12+19706d90d87784/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/kubelet/cm/devicemanager/manager.go:1035
Feb 22 05:15:41 node-01 kubelet[2837]: k8s.io/kubernetes/pkg/kubelet/cm/devicemanager.(*ManagerImpl).allocateContainerResources(0xc001157040, 0xc0012e4800, 0xc0010fb080, 0xc001da6210, 0xc001205268, 0xc0005cb300)
Feb 22 05:15:41 node-01 kubelet[2837]:         /workspace/anago-v1.19.2-rc.0.12+19706d90d87784/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/kubelet/cm/devicemanager/manager.go:844 +0x2a5 fp=0xc0005cb2c8 sp=0xc0005cb058 pc=0x1d18165
Feb 22 05:15:41 node-01 kubelet[2837]: k8s.io/kubernetes/pkg/kubelet/cm/devicemanager.(*ManagerImpl).Allocate(0xc001157040, 0xc0012e4800, 0xc0010fb080, 0x384d7affa0d5036b, 0x0)
Feb 22 05:15:41 node-01 kubelet[2837]:         /workspace/anago-v1.19.2-rc.0.12+19706d90d87784/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/kubelet/cm/devicemanager/manager.go:384 +0x2ab fp=0xc0005cb640 sp=0xc0005cb2c8 pc=0x1d126cb
Feb 22 05:15:41 node-01 kubelet[2837]: k8s.io/kubernetes/pkg/kubelet/cm/topologymanager.(*manager).allocateAlignedResources(0xc0009824c0, 0xc0012e4800, 0xc0010fb080, 0x7, 0xc0015be1e8)
Feb 22 05:15:41 node-01 kubelet[2837]:         /workspace/anago-v1.19.2-rc.0.12+19706d90d87784/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/kubelet/cm/topologymanager/topology_manager.go:187 +0x7a fp=0xc0005cb690 sp=0xc0005cb640 pc=0x1cdf67a
Feb 22 05:15:41 node-01 kubelet[2837]: k8s.io/kubernetes/pkg/kubelet/cm/topologymanager.(*manager).Admit(0xc0009824c0, 0xc001d02b40, 0x0, 0x0, 0x0, 0x0, 0x0)
Feb 22 05:15:41 node-01 kubelet[2837]:         /workspace/anago-v1.19.2-rc.0.12+19706d90d87784/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/kubelet/cm/topologymanager/topology_manager.go:264 +0x545 fp=0xc0005cb8e8 sp=0xc0005cb690 pc=0x1ce03a5
Feb 22 05:15:41 node-01 kubelet[2837]: k8s.io/kubernetes/pkg/kubelet.(*Kubelet).canAdmitPod(0xc000a57500, 0xc000bc5780, 0xc, 0x10, 0xc0012e4800, 0xc, 0x10, 0x0, 0xc000e71ea0, 0x6b4aa40)
Feb 22 05:15:41 node-01 kubelet[2837]:         /workspace/anago-v1.19.2-rc.0.12+19706d90d87784/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/kubelet/kubelet.go:1713 +0xef fp=0xc0005cb978 sp=0xc0005cb8e8 pc=0x36822ef
Feb 22 05:15:41 node-01 kubelet[2837]: k8s.io/kubernetes/pkg/kubelet.(*Kubelet).HandlePodAdditions(0xc000a57500, 0xc000c24cc0, 0x1, 0x1)
Feb 22 05:15:41 node-01 kubelet[2837]:         /workspace/anago-v1.19.2-rc.0.12+19706d90d87784/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/kubelet/kubelet.go:1987 +0x310 fp=0xc0005cba18 sp=0xc0005cb978 pc=0x3684b50
Feb 22 05:15:41 node-01 kubelet[2837]: k8s.io/kubernetes/pkg/kubelet.(*Kubelet).syncLoopIteration(0xc000a57500, 0xc0003c27e0, 0x4c946e0, 0xc000a57500, 0xc0019ba840, 0xc0019ba8a0, 0xc000cf6600, 0x1)
Feb 22 05:15:41 node-01 kubelet[2837]:         /workspace/anago-v1.19.2-rc.0.12+19706d90d87784/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/kubelet/kubelet.go:1832 +0x63f fp=0xc0005cbe40 sp=0xc0005cba18 pc=0x36830bf
Feb 22 05:15:41 node-01 kubelet[2837]: k8s.io/kubernetes/pkg/kubelet.(*Kubelet).syncLoop(0xc000a57500, 0xc0003c27e0, 0x4c946e0, 0xc000a57500)
Feb 22 05:15:41 node-01 kubelet[2837]:         /workspace/anago-v1.19.2-rc.0.12+19706d90d87784/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/kubelet/kubelet.go:1775 +0x366 fp=0xc0005cbf20 sp=0xc0005cbe40 pc=0x3682966
Feb 22 05:15:41 node-01 kubelet[2837]: k8s.io/kubernetes/pkg/kubelet.(*Kubelet).Run(0xc000a57500, 0xc0003c27e0)
Feb 22 05:15:41 node-01 kubelet[2837]:         /workspace/anago-v1.19.2-rc.0.12+19706d90d87784/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/kubelet/kubelet.go:1365 +0x2a5 fp=0xc0005cbfd0 sp=0xc0005cbf20 pc=0x367eb85
Feb 22 05:15:41 node-01 kubelet[2837]: runtime.goexit()
```
#### Environment:
- Kubernetes version (use `kubectl version`):  v1.19.2
- Cloud provider or hardware configuration:  NA
- OS (e.g: `cat /etc/os-release`):  CentOS Linux 8 (Core)
- Kernel (e.g. `uname -a`):  4.18.0-193.19.1.el8_2.x86_64

**Comment by dims** [comment]
@sseetharaman6 please see the following:
https://github.com/kubernetes/kubernetes/pull/93243/files#diff-623f1164bd7d648ee0421742187c76b5071b44315aee695f29edc0203b6c762dR1046-R1047

There's an additional mutex that was introduced in https://github.com/kubernetes/kubernetes/pull/93243 looks like you have to try 1.20.x to see if the problem is already fixed.

**Comment by sseetharaman6** [comment]
@dims, thanks I believe this is the mutex  : https://github.com/kubernetes/kubernetes/blob/v1.20.0/pkg/kubelet/cm/devicemanager/manager.go#L1051  and it is part of 1.20.x release.  Will this be backported to 1.19.x by any chance?

**Comment by dims** [comment]
@sseetharaman6 probably not as it is part of the larger https://github.com/kubernetes/kubernetes/pull/93243 which is marked a feature.

**Comment by liggitt** [comment]
the TopologyManager feature was beta in 1.18... a crashing panic seems like something a targeted fix should be backported for

**Comment by ehashman** [comment]
/remove-lifecycle stale

I think we should try to backport a targeted fix that just fixes the panic in 1.19 (1.18 unfortunately isn't supported anymore).

/triage accepted
/help

**Comment by dimakow** [comment]
Opened a PR https://github.com/kubernetes/kubernetes/pull/103262 for this bug. Please review. Happy to discuss the details!
